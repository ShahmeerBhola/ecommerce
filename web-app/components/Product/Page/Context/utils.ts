import isEmpty from 'lodash/isEmpty';

import type { TProductContext } from './types';

export const getSingleAttributes = (_attributes?: StringMap<string[]>): StringMap => {
    let singleAttributes = {};

    for (const [key, value] of Object.entries(_attributes || {})) {
        if (value.length === 1) {
            singleAttributes = {
                ...singleAttributes,
                [key]: value[0],
            };
        }
    }

    return singleAttributes;
};

export const getAvailableAttributes = (
    variations: WordPress.ProductVariation[],
    newAttributes: StringMap,
): StringMap<string[]> => {
    const availableAttribute: StringMap<string[]> = {};

    for (const { attributes } of variations) {
        let isAvailableAttribute = true;

        for (const [key, value] of Object.entries(newAttributes)) {
            isAvailableAttribute = isAvailableAttribute && attributes[key] === value;
        }

        if (isAvailableAttribute) {
            for (const [key, value] of Object.entries(attributes)) {
                const currentAttribute = availableAttribute[key] || [];

                if (currentAttribute.indexOf(value) > -1) {
                    continue;
                }

                availableAttribute[key] = [...currentAttribute, value];
            }
        }
    }

    return availableAttribute;
};

const getAttributes = (
    variations: WordPress.ProductVariation[],
    selectedAttributes: {} | StringMap,
): TProductContext.CheckAttributesType => {
    const availableAttributes = getAvailableAttributes(variations, selectedAttributes);
    const singleAttributes = getSingleAttributes(availableAttributes);

    return {
        currentAttributes: {
            ...selectedAttributes,
            ...singleAttributes,
        },
        availableAttributes,
    };
};

export const checkAttributes = (state: TProductContext.State, payload: any): TProductContext.CheckAttributesType => {
    const { currentAttributes, product } = state;
    const { variations } = product as WordPress.Product;
    const newAttributes = {
        ...currentAttributes,
        ...payload,
    };

    const attributes = getAttributes(variations, newAttributes);

    if (isEmpty(attributes.availableAttributes)) {
        const resetAttributes = {
            ...payload,
        };

        return getAttributes(variations, resetAttributes);
    }

    return attributes;
};

export const isActiveAttribute = (attributes: StringMap, currentAttributes: StringMap): boolean => {
    return Object.entries(attributes).reduce<boolean>((isSelectedAttribute, attribute) => {
        const [key, value] = attribute;
        const matchedAttribute = value === '' || currentAttributes[key] === value;

        return isSelectedAttribute && matchedAttribute;
    }, true);
};
