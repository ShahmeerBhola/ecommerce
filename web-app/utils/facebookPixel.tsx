import React from 'react';
import config from '../config';

const { facebookPixelID } = config;

// https://www.facebook.com/business/help/402791146561655?id=1205376682832142
const event = <P,>(eventName: string, params?: P): void => {
  if (facebookPixelID !== null) {
    if (params !== undefined) {
      window.fbq('track', eventName, params);
    } else {
      window.fbq('track', eventName);
    }
  }
};

export const addLineItemToCart = ({ sku, name, price }: WordPress.ProductBase, type = 'line_item'): void =>
  event<facebook.Pixel.AddToCartParameters>('AddToCart', {
    content_name: name,
    content_type: type,
    content_ids: [sku],
    value: price,
    currency: config.defaultCurrencyCode,
  });

export const addWheelTirePackageToCart = ({ wheel, tire, addOns }: Cart.WheelTirePackage): void => {
  const addWheelTirePackageItemToCart = (product: WordPress.ProductBase): void =>
    addLineItemToCart(product, 'wheel_tire_package');

  addWheelTirePackageItemToCart(wheel);
  addWheelTirePackageItemToCart(tire);
  addOns.map((addOn) => addWheelTirePackageItemToCart(addOn));
};

export const initiateCheckout = (): void => event('InitiateCheckout');
export const completePurchase = (params: facebook.Pixel.PurchaseParameters): void =>
  event<facebook.Pixel.PurchaseParameters>('Purchase', params);

export const FacebookPixelScripts: React.FunctionComponent = () => {
  {
    if (facebookPixelID !== null) {
      return (
        <>
          <script
            dangerouslySetInnerHTML={{
              __html: `
            !function(f,b,e,v,n,t,s)
{if(f.fbq)return;n=f.fbq=function(){n.callMethod ?
              n.callMethod.apply(n, arguments) : n.queue.push(arguments)};
if(!f._fbq)f._fbq=n;n.push=n;n.loaded=!0;n.version='2.0';
n.queue=[];t=b.createElement(e);t.async=!0;
t.src=v;s=b.getElementsByTagName(e)[0];
s.parentNode.insertBefore(t,s)}(window,document,'script',
'https://connect.facebook.net/en_US/fbevents.js');
 fbq('init', '${facebookPixelID}');
fbq('track', 'PageView');
          `,
            }}
          />
          <noscript>
            <img
              height="1"
              width="1"
              src={`https://www.facebook.com/tr?id=${facebookPixelID}&ev=PageView
&noscript=1`}
            />
          </noscript>
        </>
      );
    }
  }
  return null;
};
