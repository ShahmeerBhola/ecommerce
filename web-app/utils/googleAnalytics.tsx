import React from 'react';
import config from '../config';
import { isBrowserSide } from './';

const { googleAnalyticsTrackingID, googleAdsConversionID } = config;

// https://developers.google.com/analytics/devguides/collection/gtagjs/pages
// https://github.com/vercel/next.js/tree/canary/examples/with-google-analytics
export const pageview = (url: string): void => {
  if (googleAnalyticsTrackingID !== null && isBrowserSide()) {
    window.gtag('config', googleAnalyticsTrackingID, {
      page_path: url,
    });
  }
};

// https://developers.google.com/analytics/devguides/collection/gtagjs/events
const event = (action: string, eventParams: any): void => {
  if (googleAnalyticsTrackingID !== null && isBrowserSide()) {
    window.gtag('event', action, eventParams);
  }
};

const conversionEvent = (sendTo: string, eventParams?: StringMap): void =>
  event('conversion', { send_to: sendTo, ...eventParams });

export const sendOrderConfirmationConversionEvent = (orderId: string): void =>
  conversionEvent('AW-479563831/zZZcCPqr3-gBELeg1uQB', { transaction_id: orderId });

export const sendContactFormConversionEvent = (): void => conversionEvent('AW-479563831/6i-DCKyL6-gBELeg1uQB');

export const sendRequestForQuoteEvent = (): void => conversionEvent('AW-479563831/C9PsCJv7_-gBELeg1uQB');

// https://developers.google.com/analytics/devguides/collection/gtagjs/enhanced-ecommerce
export const sendEnhancedEcommerceEvent = (
  eee: GoogleAnalytics.EnhancedEcommerceEvent,
  items: GoogleAnalytics.Product[],
): void => event(eee, { items });

export const sendEnhancedEcommercePurchaseEvent = (data: GoogleAnalytics.PurchaseEvent): void =>
  event('purchase', data);

export const sendEnhancedEcommerceCheckoutOptionEvent = (data: GoogleAnalytics.CheckoutProgressEvent): void =>
  event('set_checkout_option', data);

export const enhancedEcommerceLineItemListName = 'Line Items';
export const enhancedEcommerceWheelTirePackageListName = 'Wheel Tire Package';
export const enhancedEcommerceWheelTirePackageAddOnsListName = 'Wheel Tire Package Add Ons';

export const GoogleAnalyticsAndAdsScripts: React.FunctionComponent = () => {
  {
    if (googleAnalyticsTrackingID !== null && googleAdsConversionID !== null) {
      return (
        <>
          <script async src={`https://www.googletagmanager.com/gtag/js?id=${googleAnalyticsTrackingID}`} />
          <script
            dangerouslySetInnerHTML={{
              __html: `
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', '${googleAnalyticsTrackingID}', {
              page_path: window.location.pathname,
            });
            gtag('config', '${googleAdsConversionID}')
          `,
            }}
          />
        </>
      );
    }
  }
  return null;
};
