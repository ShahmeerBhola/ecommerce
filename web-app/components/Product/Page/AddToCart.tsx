import React, { FC, HTMLAttributes, useContext } from 'react';
import { useRouter } from 'next/router';
import isEmpty from 'lodash/isEmpty';

import {
    constructUrl,
    productIsWheel,
    productIsTire,
    hasSelectedWheelQueryParam,
    productHasVariations,
} from '../../../utils';

// import { ProductVariations as ProductVariationsDropdown } from '../../components/Product';
import { CartContext } from '../../Cart';
import Button from '../../Button';
import AddToCartButton from '../../AddToCartButton';
import { TruckRequiredForProductForm, WheelTirePackageForm } from '../../Forms';

import { ProductContext } from './Context';

const ProductAddToCart: FC<HTMLAttributes<HTMLDivElement>> = (props) => {
    const router = useRouter();
    const { addLineItem } = useContext<Cart.Context>(CartContext);

    const {
        state: { product, queryParams, variation },
    } = useContext(ProductContext);

    const { id, is_purchasable, categories, truck_info_required_for_purchase } = product as WordPress.Product;
    const { year, make, model, trim } = queryParams;
    const truckFormProps = { year, make, model, trim };

    const isWheel = productIsWheel(categories);
    const isTire = productIsTire(categories);
    const hasVariations = productHasVariations(product as WordPress.Product);
    const hasActiveVariation = hasVariations && !isEmpty(variation);

    const hasSelectedWheelParam = hasSelectedWheelQueryParam(queryParams);
    const isWheelTirePackage = isTire && hasSelectedWheelParam;

    const addProductToCart = (): Promise<void> =>
        addLineItem((hasActiveVariation ? variation : product) as WordPress.Product);

    const continueToAddTires = (): void => {
        const pathname = constructUrl({ page: 'category', extra: 'tires' });
        const wheelID = !isEmpty(variation) ? variation.id : id;

        router.push({ pathname, query: { selected_wheel: wheelID } });
    };

    if (!is_purchasable) {
        return (
            <div {...props}>
                <Button className="w-full" disabled={true}>
                    Out of Stock
                </Button>
            </div>
        );
    }

    if (isWheel) {
        return (
            <div {...props}>
                <AddToCartButton
                    className="mt-2"
                    onClick={continueToAddTires}
                    inverted={false}
                    disabled={!hasActiveVariation}
                >
                    Continue to Add Tires
                </AddToCartButton>
                <AddToCartButton className="mt-2" onClick={addProductToCart} disabled={!hasActiveVariation}>
                    Buy without Tires
                </AddToCartButton>
            </div>
        );
    }

    if (truck_info_required_for_purchase) {
        return (
            <div {...props}>
                {/* {hasVariations && (
                    <ProductVariationsDropdown
                        product={product}
                        selectedVariant={product}
                        onVariantSelect={(): void => {
                            console.log('select variant');
                        }}
                        className={`mt-5 pb-2`}
                    />
                )} */}
                <TruckRequiredForProductForm
                    product={product as WordPress.ProductBase}
                    externalDisableSubmit={!hasActiveVariation}
                    {...truckFormProps}
                />
            </div>
        );
    }

    if (isWheelTirePackage) {
        return (
            <div {...props}>
                <WheelTirePackageForm tire={product as WordPress.Product} {...truckFormProps} />
            </div>
        );
    }

    // if (hasVariations) {
    //     return (
    //         <div {...props}>
    //             {/* <ProductVariations /> */}
    //             {/* <Variations product={product} /> */}
    //             <AddToCartButton className="mt-5" disabled={selectedVariantIsNull} onClick={addProductToCart} />
    //         </div>
    //     );
    // }

    return (
        <div {...props}>
            <AddToCartButton className="mt-5" disabled={!hasActiveVariation} onClick={addProductToCart} />
        </div>
    );
};

export default ProductAddToCart;
