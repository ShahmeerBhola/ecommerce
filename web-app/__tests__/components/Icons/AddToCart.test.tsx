import React from 'react';
import renderer from 'react-test-renderer';
import { generateMockProduct } from '../../utils';
import AddToCart from '../../../components/Icons/AddToCart';

describe('AddToCart component', () => {
  test('it renders properly', () => {
    const tree = renderer.create(<AddToCart product={generateMockProduct()} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  // TODO: test that when it is clicked it calls addLineItem context method
});
