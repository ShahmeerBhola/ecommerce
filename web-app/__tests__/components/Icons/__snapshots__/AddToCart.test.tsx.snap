// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddToCart component it renders properly 1`] = `
<div
  className="cursor-pointer inline-block"
  onClick={[Function]}
>
  <svg
    fill="none"
    height="50"
    viewBox="0 0 50 50"
    width="50"
    xmlns="http://www.w3.org/2000/svg"
  >
    <rect
      fill="#F23939"
      height="50"
      width="50"
    />
    <path
      d="M18.325 32.5875H30.3661L31.1103 29.5588H15.3743L12 15.8266H34.4845L35.6705 11H42V12.7979H37.0503L32.1699 32.6593C33.5906 32.9697 34.6591 34.2569 34.6591 35.7938C34.6591 37.5617 33.2454 39 31.5077 39C29.77 39 28.3563 37.5616 28.3563 35.7938C28.3563 35.2887 28.472 34.8108 28.6774 34.3854H21.1553C21.3607 34.8108 21.4764 35.2887 21.4764 35.7938C21.4764 37.5617 20.0627 39 18.325 39C16.5874 39 15.1736 37.5616 15.1736 35.7938C15.1737 34.0259 16.5874 32.5875 18.325 32.5875Z"
      fill="white"
    />
    <path
      d="M23.7656 19H26.2656V27H23.7656V19Z"
      fill="#F23939"
    />
    <path
      d="M29 21.7383L29 24.2383L21 24.2383L21 21.7383L29 21.7383Z"
      fill="#F23939"
    />
  </svg>
</div>
`;
